nanomq/apps/webhook_test.c

```diff
@@ -0,0 +1,478 @@
+	nanomq_conf->daemon                 = false;
+	nanomq_conf->num_taskq_thread       = 1;
+	nanomq_conf->max_taskq_thread       = 1;
+	nanomq_conf->parallel               = 10;
+	nanomq_conf->property_size          = 32;
+	nanomq_conf->max_packet_size        = 1024;
+	nanomq_conf->client_max_packet_size = 1024;
+	nanomq_conf->msq_len                = 32;
+	nanomq_conf->qos_duration           = 2;
+
+	nanomq_conf->sqlite.enable      = false;
+	nanomq_conf->tls.enable         = false;
+	nanomq_conf->websocket.enable   = false;
+	nanomq_conf->http_server.enable = false;
+
+	// conf for webhook
+	nanomq_conf->web_hook.enable         = true;
+	nanomq_conf->web_hook.url            = "http://0.0.0.0:8888/hook";
+	nanomq_conf->web_hook.encode_payload = plain;
+	nanomq_conf->web_hook.pool_size      = 32;
+
+	// set up webhook headers
+	nanomq_conf->web_hook.header_count = 1;
+	nanomq_conf->web_hook.headers = realloc(nanomq_conf->web_hook.headers,
+	    nanomq_conf->web_hook.header_count * sizeof(conf_http_header *));
+		
+	header                        = calloc(1, sizeof(conf_http_header));
+	header->key                   = "content-type";
+	header->value                 = "application/json";
+	nanomq_conf->web_hook.headers[0] = header;
+
+	// set up webhook rules
+	nanomq_conf->web_hook.rule_count = 5;
+	nanomq_conf->web_hook.rules      = realloc(nanomq_conf->web_hook.rules,
+	         nanomq_conf->web_hook.rule_count * sizeof(conf_web_hook_rule *));
+
+	webhook_rule                   = calloc(1, sizeof(conf_web_hook_rule));
+	webhook_rule->event            = CLIENT_CONNECT;
+	webhook_rule->rule_num         = 1;
+	webhook_rule->action           = "on_client_connect";
+	nanomq_conf->web_hook.rules[0] = webhook_rule;
+	webhook_rule                   = calloc(1, sizeof(conf_web_hook_rule));
+	webhook_rule->event            = CLIENT_CONNACK;
+	webhook_rule->rule_num         = 1;
+	webhook_rule->action           = "on_client_connack";
+	nanomq_conf->web_hook.rules[1] = webhook_rule;
+	webhook_rule                   = calloc(1, sizeof(conf_web_hook_rule));
+	webhook_rule->event            = CLIENT_CONNECTED;
+	webhook_rule->rule_num         = 1;
+	webhook_rule->action           = "on_client_connected";
+	nanomq_conf->web_hook.rules[2] = webhook_rule;
+	webhook_rule                   = calloc(1, sizeof(conf_web_hook_rule));
+	webhook_rule->event            = CLIENT_DISCONNECTED;
+	webhook_rule->rule_num         = 1;
+	webhook_rule->action           = "on_client_disconnected";
+	nanomq_conf->web_hook.rules[3] = webhook_rule;
+	webhook_rule                   = calloc(1, sizeof(conf_web_hook_rule));
+	webhook_rule->event            = MESSAGE_PUBLISH;
+	webhook_rule->rule_num         = 1;
+	webhook_rule->action           = "on_message_publish";
+	nanomq_conf->web_hook.rules[4] = webhook_rule;
+
+	return nanomq_conf;
+}
+
+int
+main(int argc, char **argv)
+{
+	char *cmd_pub = "mosquitto_pub -h 127.0.0.1 -p 1881 -t topic1 -m message -q 2";
```

Comment:
It would be better to use a macro rather than a fixed number as the port.

Reply:
same reason as mentioned above.